File: user-service/src/main/java/com/phorus/userservice/model/dtos/UserDTO.java
package com.phorus.userservice.model.dtos;

import com.phorus.userservice.model.validationGroups.UserValidationGroup;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;

import javax.validation.constraints.*;


@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@SuperBuilder
public class UserDTO {

    @NotBlank(groups = UserValidationGroup.UserCreateValGroup.class, message = "Cannot be blank")
    @Size(groups = {UserValidationGroup.UserCreateValGroup.class, UserValidationGroup.UserUpdateValGroup.class},
            min = 3,
            message = "Must have at least 3 characters long")
    private String name;

    @NotNull(groups = UserValidationGroup.UserCreateValGroup.class, message = "Cannot be null")
    @Min(groups = {UserValidationGroup.UserCreateValGroup.class, UserValidationGroup.UserUpdateValGroup.class},
            value = 0,
            message = "Cannot be negative")
    private Integer age;
}

File: user-service/src/main/kotlin/com/phorus/userservice/model/dtos/UserDTO.kt
package com.phorus.userservice.model.dtos

import com.phorus.userservice.model.validationGroups.UserCreateValGroup
import com.phorus.userservice.model.validationGroups.UserUpdateValGroup

import javax.validation.constraints.Min
import javax.validation.constraints.NotBlank
import javax.validation.constraints.NotNull
import javax.validation.constraints.Size

data class UserDTO(
    @field:NotBlank(groups = [UserCreateValGroup::class], message = "Cannot be blank")
    @field:Size(
        groups = [UserCreateValGroup::class, UserUpdateValGroup::class],
        min = 3,
        message = "Must have at least 3 characters long"
    )
    var name: String? = null,

    @field:NotNull(groups = [UserCreateValGroup::class], message = "Cannot be null")
    @field:Min(
        groups = [UserCreateValGroup::class, UserUpdateValGroup::class],
        value = 0,
        message = "Cannot be negative"
    )
    var age: Int? = null,
)

data class UserResponse(
    val id: Long,
    val name: String,
    val age: Int,
)